packages/protoplugin/src/ecmascript/custom-options.ts

```diff
@@ -0,0 +1,367 @@
+// Copyright 2021-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { AnyDesc } from "@bufbuild/protobuf";
+import {
+  proto3,
+  BinaryReader,
+  WireType,
+  MessageType,
+  AnyMessage,
+  ScalarType,
+} from "@bufbuild/protobuf";
+
+/**
+ * Returns a binary reader for the given descriptor and field ID.
+ */
+function getBinaryReader(
+  desc: AnyDesc,
+  id: number,
+  expectedWireType: WireType
+): BinaryReader | undefined {
+  const opt = desc.proto.options;
+  let reader: BinaryReader | undefined = undefined;
+  if (opt !== undefined) {
+    const unknownFields = proto3.bin.listUnknownFields(opt);
+    const field = unknownFields.find((f) => f.no === id);
+    if (field) {
+      if (field.wireType !== expectedWireType) {
+        throw new Error(
+          `expected wire type ${expectedWireType}, but got ${field.wireType}`
+        );
+      }
+      reader = new BinaryReader(field.data);
+    }
+  }
+  return reader;
+}
+
+type ScalarValue = number | bigint | string | undefined;
+
+export function getCustomScalarOption(
```

Comment:
Put this as an example for maybe having a single function for scalar values instead of one per type.

Reply:
Makes a lot of sense to me! 

If we deliberately don't support repeated fields (which is reasonable), a function each for scalar, message, and enum fields seems like a decent API.

What do you think about letting it return a `string` for `ScalarType.STRING`, a `boolean` for `BOOL` etc?